Automattic/wp-calypso
# build: Clean up TypeScript config

#### Changes proposed in this Pull Request

Clean up `./client/tsconfig.json` to make it more consistent with other TypeScript projects in `./packages/`:

* Create a new shared `mixed-package.json` meant to be used by packages that have TypeScript and JavaScript. Use it as the base for `./client/tsconfig.json`
* Use a declaration file to import the types from `@automattic/calypso-build`. The old `typeRoots` forces us to use relative paths (problematic if the package is moved in the future) and re-declare `node_modules/@types`. Having it in a declaration file allows us to import from the package name `@automattic/calypso-build`.

The new config can be tested with `yarn tsc --build client`. It should be reasonably fast (47 sec on my MBP 2017, 25 sec if `client/tsconfig.tsbuildinfo` already exists), but will throw a ton of errors though. Until we fix the errors, we can't easily make it part of `yarn build-packages` or the `postinstall` script (and it won't be a huge benefit anyway)

The next step is to run `yarn tsc --build client` as part of CI, capture the errors and expose them as Code Style issues. That way we can progressively fix them and track the progress, like we do with ESLint errors.

I also did a few fixes to problems that were detected by the new config file:
* We had a package-relative import `state/automated-transfer/selectors` that were not prefixed by `calypso/`
* Old import from `../../../config` instead of the extracted `@automattic/calypso-config`
* Missing `@types/store`
* Empty `.ts` file that TypeScript detects as a global script and can't be compiled using `--isolatedModules`


#### Testing instructions

Run `time yarn tsc --build client`, you should get many errors (591) and it should be reasonably quick (46s for me). Run it a second time, it should be about twice as fast.
